worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/IngestionBatchWorkerPlugin.java

```diff
@@ -0,0 +1,272 @@
+    }
+
+    @Override
+    public void processBatch(final BatchWorkerServices batchWorkerServices, final String batchDefinition,
+                             final String taskMessageType, final Map<String, String> taskMessageParams)
+        throws BatchDefinitionException
+    {
+        log.debug("Definitions received: " + batchDefinition);
+        if (taskMessageParams != null) {
+            log.debug("Task message params: " + taskMessageParams.toString());
+        }
+
+        if (batchDefinition == null || batchDefinition.trim().length() == 0) {
+            log.error("IngestionBatchWorkerPlugin has not received a valid batch definition string");
+            throw new RuntimeException("IngestionBatchWorkerPlugin has not received a valid batch definition string");
+        }
+        if (batchDefinition.startsWith("subbatch:")) {
+            try {
+                // this is already a subbatch
+                // read the file from the staging and send it through
+                log.debug("Subbatch received: " + batchDefinition);
+                handleSubbatch(batchDefinition, batchWorkerServices, taskMessageParams);
+            } catch (InvalidBatchIdException ex) {
+                log.error("Invalid Batch Id Exception! " + ex.getMessage());
+                throw new RuntimeException("Invalid Batch Id Exception! " + ex.getMessage());
+            }
+        } else if (batchDefinition.contains("|")) {
+            log.debug("Multiple Batches received: " + batchDefinition);
+            handleMultipleBatchIds(batchDefinition, batchWorkerServices);
+        } else {
+            log.debug("Single Batch received: " + batchDefinition);
+            handleSingleBatchId(batchDefinition, batchWorkerServices);
+        }
+
+    }
+
+    private void handleMultipleBatchIds(final String batchIds, final BatchWorkerServices batchWorkerServices)
+    {
+        final String tenantId = extractTenantId(batchIds);
+        final String[] batchesSplit = extractBatchIds(batchIds);
+        Arrays.stream(batchesSplit).parallel().forEach(batch -> batchWorkerServices.registerBatchSubtask(tenantId + "/" + batch));
+    }
+
+    private void handleSingleBatchId(final String batchId, final BatchWorkerServices batchWorkerServices)
+    {
+        try {
+            final TenantId tenantId = new TenantId(extractTenantId(batchId));
+            final BatchId batchIdExtracted = new BatchId(extractBatchIds(batchId)[0]);
+            final Path pathOfSubBatches = fileSystemProvider.getPathForBatch(tenantId, batchIdExtracted);
+            final String[] extensions = {"batch"};
+            final Collection<File> subbatchesFiles = FileUtils.listFiles(pathOfSubBatches.toFile(), extensions, false);
+            subbatchesFiles.parallelStream().forEach(f -> log.debug("Batch file found: " + FilenameUtils.getName(f.getAbsolutePath())));
+            subbatchesFiles
+                .parallelStream()
+                .forEach(subbatch
```

Comment:
As discussed, don't kick off other threads to callback into the worker framework.  It's not clear to me that that is safe, particularly since you are returning from the call.